File: problems/树/129.求根到叶子数字之和.cpp
#include <iostream>
#include <functional>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int sumNumbers(TreeNode* root) {
    int result = 0;
    function<void(TreeNode*, int)> dfs = [&](TreeNode* node, int num) {
        if (!node) return;
        num = num * 10 + node->val;
        if (!node->left && !node->right) { result += num; return; }
        dfs(node->left, num);
        dfs(node->right, num);
    };
    dfs(root, 0);
    return result;
}
int main() {
    TreeNode* root = new TreeNode(1);
    root->left  = new TreeNode(2);
    root->right = new TreeNode(3);
    cout << sumNumbers(root) << endl;             // 25
    cout << sumNumbers(new TreeNode(5)) << endl;   // 5
    cout << sumNumbers(nullptr) << endl;           // 0
    return 0;
}

File: problems/树/129.求根到叶子数字之和.py
#!/usr/bin/env python3
"""
129. 求根到叶子数字之和
"""
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def sum_numbers(root: Optional[TreeNode]) -> int:
    result = 0
    def dfs(node: Optional[TreeNode], num: int):
        nonlocal result
        if not node: return
        num = num * 10 + node.val
        if not node.left and not node.right: result += num
        dfs(node.left, num)
        dfs(node.right, num)
    dfs(root, 0)
    return result

if __name__ == "__main__":
    root1 = TreeNode(1, TreeNode(2), TreeNode(3))
    print(sum_numbers(root1))  # 25
    print(sum_numbers(TreeNode(5)))  # 5
    print(sum_numbers(None))  # 0
